src/optimisation/Optimisation.jl

```diff
@@ -1,324 +1,327 @@
+generated either by sampling from the prior (if no constraints are present) or uniformly
+from the box constraints. If generic constraints are set, an error is thrown.
+"""
+function generate_init_value(model::Model, init_value, constraints)
+    if init_value !== nothing
+        return copy(init_value)
+    end
+    if has_generic_constraints(constraints)
+        throw(ArgumentError(
+            "You must provide an initial value when using generic constraints."
+        ))
     end
-    return r, lp, vi
+    if has_box_constraints(constraints)
+        return [
+            rand(Uniform(lower, upper))
+            for (lower, upper) in zip(constraints.lb, constraints.ub)
+        ]
+    end
+    return rand(model) |> values |> Iterators.flatten |> collect
 end
 
 """
-    OptimLogDensity{M<:Model,C<:Context,V<:VarInfo}
+    ModeEstimationProblem{M,E,Iv,Cons,OLD,ADType}
 
-A struct that stores the negative log density function of a `DynamicPPL` model.
-"""
-const OptimLogDensity{M<:Model,C<:OptimizationContext,V<:VarInfo} = Turing.LogDensityFunction{V,M,C}
+A struct that defines a mode estimation problem.
 
+# Fields
+$(TYPEDFIELDS)
 """
-    OptimLogDensity(model::Model, context::OptimizationContext)
-
-Create a callable `OptimLogDensity` struct that evaluates a model using the given `context`.
-"""
-function OptimLogDensity(model::Model, context::OptimizationContext)
-    init = VarInfo(model)
-    return Turing.LogDensityFunction(init, model, context)
+struct ModeEstimationProblem{
+    M<:Model,
+    E<:ModeEstimator,
+    Iv<:AbstractVector,
+    Cons<:ModeEstimationConstraints,
+    OLD<:OptimLogDensity,
+    ADType<:AbstractADType,
+}
+    "The DynamicPPL model for which to estimate the mode."
+    model::M
+    """The mode estimator. Either `MLE()` for maximum likelihood or `MAP()` for maximum a
+    posteriori."""
+    estimator::E
+    "The initial value for the optimization."
+    init_value::Iv
+    "The constraints for the optimization problem."
+    constraints::Cons
+    "`OptimLogDensity` of the model, used as the objective function for the optimization."
+    log_density::OLD
+    "The automatic differentiation type to use. See ADTypes.jl."
+    adtype::ADType
+    "Whether the objective function is transformed to an unconstrained space."
+    linked::Bool
 end
 
 """
-    LogDensityProblems.logdensity(f::OptimLogDensity, z)
-
-Evaluate the negative log joint (with `DefaultContext`) or log likelihood (with `LikelihoodContext`)
-at the array `z`.
-"""
-function (f::OptimLogDensity)(z::AbstractVector)
-    varinfo = DynamicPPL.unflatten(f.varinfo, z)
-    return -getlogp(last(DynamicPPL.evaluate!!(f.model, varinfo, f.context)))
-end
-
-# NOTE: This seems a bit weird IMO since this is the _negative_ log-likelihood.
-LogDensityProblems.logdensity(f::OptimLogDensity, z::AbstractVector) = f(z)
+    ModeEstimationProblem(model, estimator, init_value, lb, ub, cons, lcons, ucons, adtype)
```

Comment:
Generally want to keep method signatures to one line. If it overflows, just do `ModeEstimationProblem(args...)` and explain the args below.

Reply:
Done